lib/mdl/rules.rb

```diff
@@ -796,3 +796,85 @@ def check_blockquote(errors, elements)
     error_lines
   end
 end
+
+rule 'MD055', 'Tables should be well-formed' do
+  tags :tables
+  aliases 'table-format-validation'
+  check do |doc|
+    error_lines = []
+    tables = doc.find_type_elements(:table)
+    lines = doc.lines
+    num_lines = lines.length
+
+    tables.each do |table|
+      loc = table.options[:location]-1
+      header = lines[loc]
+
+      if header[0] != '|' || header[-1] != '|' || header.length == 1
+        # Tables should start and end with '|'
+        error_lines << (loc+1)
+      end
+
+      headings = header.strip().split('|').select{ |x| x.length > 0 }
+      num_headings = headings.length
+
+      if loc+1 < num_lines
+        # Second row should be in the form | ---- | ---- |
+        second_row = lines[loc+1]
+        columns = second_row.strip().split('|').select{ |x| x.length > 0 }
+        if num_headings != columns.length
+          # Number of headings do not match with the second row
+          error_lines << (loc+2)
+        end
+        
+        # This pattern matches if
+        #   1) The row starts and stops with | characters
+        #   2) Only consists of characters '|', '-', ':' and whitespace
+        #   3) Each section between the seperators (i.e. '|')
+        #      a) has at least three consecutive dashes
+        #      b) can have whitespace at the beginning or the end
+        #      c) can have colon before and/or after the dashes (for alignment purposes)
+        #   Some examples:
+        #       |-----|----|-------|      --> matches
+        #       |:---:|:---|-------|      --> matches
+        #       |  :------:  | ----|      --> matches
+        #       | - - - | - - - |         --> does NOT match
+        #       |::---|                   --> does NOT match
+        #       |----:|:--|----|          --> does NOT match
+        pattern = /^(\|\s*:?-{3,}:?\s*)+\|$/
+        unless second_row.match(pattern)
+          # Second row is not in the form described by the pattern
+          error_lines << (loc+2)
+        end
+
+        # Check next lines until the end of the table
+        line_num = loc+2
+        while second_row.strip() != '' && line_num < num_lines
+          line = lines[line_num]
+          stripped_line = line.strip()
+
+          # If the line has '|' character, it's "possibly" a table row
+          # because the previous line was also a table row
+          if line.include?('|')
```

Comment:
is there a need to check this against line instead of stripped_line? It feels like we can probably just do like `line = lines[line_num].strip` or so?

Reply:
oh I guess you only really want that for counting columns, and not the rest?

Reply:
I was checking the pipe character to determine where to stop processing lines as part of the table in question. 

This logic won't pick up the last rows if there are blank lines in the table. But, Kramdown picks up the second portion as a different table object anyway. So, it'll fail for the "second row" condition.
```
| Heading 1 | Heading 2 |
| ---------- |-----------|
|    Row        |      Row     |

|    Row        |      Row     | 
```

I was trying to eliminate the case
```
| Heading 1 | Heading 2 |
| ---------- |-----------|
|    Row        |      Row     |
This line is not part of the table
This line is not part of the table
This line is not part of the table
This line is not part of the table
This line is not part of the table
```

Reply:
oh I get the `include?('|')` part. I was just wasn't sure why you were keeping around both `line` and `stripped_line`, which it seems like one would do for everything.

Reply:
Oh, sorry, I wasn't reading carefully. So, yes, I could use `stripped_line` for this check. Inside this part, the line is used to detect if the complete line starts and ends with pipe. I used `stripped_line` for column counting, ignoring the starting and ending spaces. Seems like with more rules instead of one, it'll be easier to decouple this logic.